packages/gatsby/src/utils/start-server.ts

```diff
@@ -495,7 +495,7 @@ module.exports = {
             // Let renderDevHTML figure it out.
             page: undefined,
             store,
-            htmlComponentRendererPath: `${program.directory}/public/render-page.js`,
+            htmlComponentRendererPath: pageRenderer,
```

Comment:
I think `pageRenderer` will be undefined here for the `else` codepath on [lines 147-153](https://github.com/gatsbyjs/gatsby/pull/28648/files#diff-c2143e514553a30526173cdb673a058ca6cbdb8f6b8d8d3e8a47edd2bce58051L147-L153):

```js
 let pageRenderer: string
  if (process.env.GATSBY_EXPERIMENTAL_DEV_SSR) {
    // ...
    pageRenderer = await buildRenderer(program, Stage.DevelopHTML)
    // ...
  } else {
    // we don't set `pageRenderer` value in this codepath
```

Reply:
It should be in the same process.env.GATSBY_EXPERIMENTAL_DEV_SSR, right? I've tested it and didn't have an issue.

Reply:
Oh, true. Didn't see the this `if`. Still surprised TS doesn't complain about this. But it's probably because `renderDevHTML` is not fully typed.

Reply:
good point :joy: